### PR TITLE
Separating Code Generation Step 2: Build Manager Service Triggers CodeBuild

### PR DESCRIPTION
**Pull request by EugeneTseitlin**
Issue Number: #3082 

## PR Details

Created Build Manager Service that received message from Kafka and based on that data triggers job in CodeBuild.

## PR Checklist
- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/code_of_conduct.md) file for detailed contributing guidelines.

### CODE REVIEW

**Review by arielweinberger — CHANGES_REQUESTED** [review]
Looks really good overall, left a few important comments. 

It would be great to get a `README.md` file explaining the relation between this service as a microservice, and the generator SDK/standalone node lib.

Well done! ⭐️

File: packages/amplication-build-manager-service/buildspec.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+version: 0.2
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/buildspec.yaml:1]: Done

File: packages/amplication-build-manager-service/src/app.controller.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { GenerateResource } from './codeBuild/dto/GenerateResource';
+
+@Controller()
+export class AppController {
+  constructor(
+    private readonly codeBuildService: CodeBuildService,
+    private readonly buildContextStorage: BuildContextStorageService,
+  ) {}
+
+  @MessagePattern('build.internal.generate-resource.event.0')
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.controller.ts:14]: Done

File: packages/amplication-build-manager-service/src/app.controller.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  async receiveCodeGenRequest(@Payload() message: any) {
+    const gr: GenerateResource = message.value;
+
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.controller.ts:17]: Done

File: packages/amplication-build-manager-service/src/app.module.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    ClientsModule.register([
+      {
+        name: 'MAIN_KAFKA',
+        transport: Transport.KAFKA,
+        options: {
+          client: {
+            brokers: ['localhost:9092'],
+          },
+          consumer: {
+            //TODO: Expose groupId via env
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.module.ts:21]: Done

File: packages/amplication-build-manager-service/src/app.module.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+@Module({
+  imports: [
+    ConfigModule.forRoot({ isGlobal: true }),
+    ClientsModule.register([
+      {
+        name: 'MAIN_KAFKA',
+        transport: Transport.KAFKA,
+        options: {
+          client: {
+            brokers: ['localhost:9092'],
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.module.ts:18]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.module.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { Module } from '@nestjs/common';
+import { BuildContextStorageService } from './buildContextStorage.service';
+
+@Module({
+  imports: [],
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.module.ts:5]: Done

File: packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import { ConfigService } from '@nestjs/config';
+import { CODE_BUILD_PROJECT_NAME } from 'src/constants';
+
+@Injectable()
+export class CodeBuildService {
+  private readonly codeBuildClient: CodeBuildClient;
+  private readonly projectName: string;
+
+  constructor(private readonly configService: ConfigService) {
+    //TODO: Provide credentials
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts:16]: Done

File: packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    this.codeBuildClient = new CodeBuildClient({
+      // credentials: {
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts:18]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  private readonly BUILD_CONTEXT_S3_LOCATION: string;
+
+  constructor(private readonly configService: ConfigService) {
+    this.client = new S3Client({
+      credentials: {
+        accessKeyId: configService.get('AWS_KEY_ID'),
+        secretAccessKey: configService.get('AWS_SECRET_KEY'),
+      },
+    });
+
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:27]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+} from '@aws-sdk/client-s3';
+import { promises as fsPr } from 'fs';
+import { GenerateResource } from 'src/codeBuild/dto/GenerateResource';
+import JSZip from 'jszip';
+import { join } from 'path';
+
+@Injectable()
+export class BuildContextStorageService {
+  private readonly client: S3Client;
+
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:16]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
+import { Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import {
+  S3Client,
+  PutObjectCommand,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { promises as fsPr } from 'fs';
+import { GenerateResource } from 'src/codeBuild/dto/GenerateResource';
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:9]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+      `${genResource.buildId}.zip`,
+    );
+
+    const uploadParams: PutObjectCommandInput = {
+      Bucket: this.BUILD_CONTEXT_S3_BUCKET,
+      Key: key,
+      Body: archive,
+    };
+
+    await this.client.send(new PutObjectCommand(uploadParams));
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:60]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      'BUILD_CONTEXT_S3_LOCATION',
+    );
+  }
+
+  public async saveBuildContextSource(
+    genResource: GenerateResource,
+  ): Promise<string> {
+    const buffer = await fsPr.readFile(genResource.contextFileLocation.path);
+
+    const zipper = new JSZip();
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:41]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:60]: Done

File: packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      //   accessKeyId: configService.get('AWS_KEY_ID'),
+      //   secretAccessKey: configService.get('AWS_SECRET_KEY'),
+      // },
+    });
+    this.projectName = this.configService.get<string>(CODE_BUILD_PROJECT_NAME);
+  }
+
+  async runBuild(contextArchivePath: string) {
+    const sbci: StartBuildCommandInput = {
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts:27]: Done

File: packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+      projectName: this.projectName,
+      sourceLocationOverride: contextArchivePath,
+    };
+
+    const sbc = new StartBuildCommand(sbci);
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts:32]: Done

File: packages/amplication-build-manager-service/src/codeBuild/dto/FileLocation.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import { StorageTypeEnum } from './StorageTypeEnum';
+
+export interface FileLocation {
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/dto/FileLocation.ts:3]: @mshidlov It's a sugestion to update contract between services. For me it sounds reasonable. What do you think? If you're agree, please, reflect the change in the design.

File: packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+export enum StorageTypeEnum {
+  FS,
+  S3,
+  GCS,
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts:4]: @mshidlov Please, confirm that GCS is not going to get back and we can remove it from a contract. Also reflect the change in the design.

File: packages/amplication-data-service-generator/src/index.ts [hunk 18]
```diff
@@ -1,2 +1,49 @@
...
+import { BuildContext, Module } from "./types";
+
+const [, , source, destination] = process.argv;
+if (!source) {
+  throw new Error("INPUT is not defined");
+}
+if (!destination) {
+  throw new Error("OUTPUT is not defined");
+}
+generateCode(source, destination).catch((error) => {
```
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-data-service-generator/src/index.ts:14]: Done

File: packages/amplication-build-manager-service/src/app.controller.ts [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+import { GenerateResource } from './codeBuild/dto/GenerateResource';
+
+@Controller()
+export class AppController {
+  constructor(
+    private readonly codeBuildService: CodeBuildService,
+    private readonly buildContextStorage: BuildContextStorageService,
+  ) {}
+
+  @EventPattern('build.internal.generate-resource.event.0')
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/app.controller.ts:14]: move topic name to env

File: packages/amplication-build-manager-service/src/app.controller.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  async receiveCodeGenRequest(@Payload() message: any) {
+    const gr: GenerateResource = message.value;
+    const path = await this.buildContextStorage.saveBuildContextSource(gr);
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/app.controller.ts:17]: add error handling

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { promises as fsPr } from 'fs';
+import { GenerateResource } from '../codeBuild/dto/GenerateResource';
+import JSZip from 'jszip';
+import { join } from 'path';
+
+@Injectable()
+export class BuildContextStorageService {
+  private readonly storageClient: S3Client;
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:15]: Add abstraction to storage

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    );
+  }
+
+  public async saveBuildContextSource(
+    genResource: GenerateResource,
+  ): Promise<string> {
+    try {
+      const buffer = await fsPr.readFile(genResource.contextFileLocation.path);
+
+      const zipper = new JSZip();
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:40]: Create compression service

File: packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (arielweinberger) [reply packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts:4]: I don't see any GCS related code in this PR, so whether it'll be back or not doesn't matter for the review.

If and when we add GCS functionality, we can add this back.

File: packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts [hunk 23]
```diff
@@ -0,0 +1,5 @@
+export enum StorageTypeEnum {
+  FS,
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts:2]: Change enum keys from integer to string

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+      zipper.file('input-data.json', buffer);
+      const archive = await zipper.generateAsync({ type: 'uint8array' });
+
+      const date = new Date().toISOString().split('T')[0];
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:44]: move it to util

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+      const key = join(
+        this.BUILD_CONTEXT_S3_LOCATION,
+        date,
+        genResource.projectId,
+        genResource.resourceId,
+        `${genResource.buildId}.zip`,
+      );
+
+      const uploadParams: PutObjectCommandInput = {
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:53]: Move it to S3StorageClient

File: packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+  CodeBuildClient,
+  StartBuildCommand,
+  StartBuildCommandInput,
+} from '@aws-sdk/client-codebuild';
+import { Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { CODE_BUILD_PROJECT_NAME } from 'src/constants';
+
+@Injectable()
+export class CodeBuildService {
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts:11]: Add abstraction to BuildService

File: packages/amplication-build-manager-service/src/app.controller.ts (hunk 19 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.controller.ts:14]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts (hunk 21 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:15]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts (hunk 22 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:40]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts (hunk 24 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:44]: Done

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts (hunk 25 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:53]: Done

File: packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts (hunk 26 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/codeBuild.service.ts:11]: Done

File: packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts (hunk 23 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts:2]: This change has to be done in a separate PR.

File: packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/dto/StorageTypeEnum.ts:4]: This change has to be done in a separate PR.

File: packages/amplication-build-manager-service/src/codeBuild/dto/FileLocation.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/codeBuild/dto/FileLocation.ts:3]: This change has to be done in a separate PR.

File: package.json [hunk 27]
```diff
@@ -19,6 +19,7 @@
     "generate": "run-s generate-graphql-schema generate-graphql-code",
     "build-storybook": "lerna run build-storybook",
     "generate-test-data-service": "lerna run --scope @amplication/data-service-generator generate-test-data-service",
+    "generate-code": "lerna run --scope @amplication/data-service-generator generate-code",
```
Comment (mshidlov) [inline package.json]: this is not part of this pr

File: packages/amplication-build-manager-service/src/app.controller.ts [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+  @EventPattern(
+    EnvironmentVariables.instance.get(GENERATE_RESOURCE_TOPIC, true),
+  )
+  async receiveCodeGenRequest(@Payload() message: any) {
+    try {
+      const gr: GenerateResource = message.value;
+      const path = await this.buildContextStorage.saveBuildContextSource(gr);
+      await this.buildService.runBuild(path);
+    } catch (error) {
+      console.error(error);
```
Comment (mshidlov) [inline packages/amplication-build-manager-service/src/app.controller.ts:26]: message logger + log + send build Status message

Comment (mshidlov) [reply packages/amplication-build-manager-service/src/app.controller.ts:26]: BuildStaus message with status Fail + message error message

File: packages/amplication-build-manager-service/src/app.controller.ts (hunk 20 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.controller.ts]: Done

File: package.json (hunk 27 above)
Comment (EugeneTseitlin) [reply package.json]: Removed

File: packages/amplication-build-manager-service/src/app.controller.ts (hunk 28 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.controller.ts:46]: Done

File: packages/amplication-build-manager-service/src/app.controller.ts [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+    this.buildStatusTopic = this.configService.get<string>(BUILD_STATUS_TOPIC);
+  }
+
+  @EventPattern(
+    EnvironmentVariables.instance.get(GENERATE_RESOURCE_TOPIC, true),
+  )
+  async receiveCodeGenRequest(@Payload() message: any) {
+    try {
+      const gr: GenerateResource = message.value;
+      const path = await this.buildContextStorage.saveBuildContextSource(gr);
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/app.controller.ts:33]: Check if we have to zip source for code build

File: packages/amplication-build-manager-service/package.json [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+    "@nestjs/config": "1.1.5",
+    "@nestjs/core": "8.2.3",
+    "@nestjs/microservices": "^8.4.5",
+    "@nestjs/platform-express": "^8.0.0",
+    "jszip": "^3.10.0",
+    "reflect-metadata": "^0.1.13",
+    "rimraf": "^3.0.2"
+  },
+  "devDependencies": {
+    "@nestjs/cli": "^8.0.0",
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/package.json:37]: lock versions in package json

File: packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+      await this.storageService.saveFile(
+        this.BUILD_CONTEXT_S3_BUCKET,
+        key,
+        archive,
+      );
+
+      const path = join(this.BUILD_CONTEXT_S3_BUCKET, key);
+      return path;
+    } catch (err) {
+      console.error(err);
```
Comment (EugeneTseitlin) [inline packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:60]: don't use log and throw after that

Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts:60]: When our logger is ready get back and use it here

Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/buildContextStorage/buildContextStorage.service.ts]: Done

File: packages/amplication-build-manager-service/src/app.controller.ts (hunk 29 above)
Comment (EugeneTseitlin) [reply packages/amplication-build-manager-service/src/app.controller.ts:36]: Yeah, we have to

**Review by arielweinberger — APPROVED** [review]

File: packages/amplication-build-manager-service/package.json (hunk 30 above)
Comment (arielweinberger) [reply packages/amplication-build-manager-service/package.json:39]: Your call what you want to do with this. I understand the need for consistency and version pinning, on the other hand I understand your argument about manual work because if we pin we might want to pin to the latest minor/patch version, and that needs to be done manually. Time is precious.
